### PR TITLE
Optimise XmlNodeConverter

### PR DESCRIPTION
**Pull request by JonHanna**
Short-circuit `SerializeGroupedNodes` on no child nodes: Avoids creating unused `Dictionary`.

Don't create `Dictionary` in `ReadAttributeElements` if it'll be empty.

Reduce calls to `JsonReader.TokenType` virtual property.

### CODE REVIEW

File: Src/Newtonsoft.Json/Converters/XmlNodeConverter.cs [hunk 1]
```diff
@@ -1177,6 +1177,11 @@ private bool IsArray(IXmlNode node)
 
         private void SerializeGroupedNodes(JsonWriter writer, IXmlNode node, XmlNamespaceManager manager, bool writePropertyName)
         {
+            if (node.ChildNodes.Count == 0)
```
Comment (JamesNK) [inline Src/Newtonsoft.Json/Converters/XmlNodeConverter.cs]: I ran this through a memory profiler and a collection is still allocated even though it will always be empty. Using XmlNode.HasChildNodes avoids allocating the ChildNodes collection. There will be other places this could be optimized.

To add this to the wrapper implementation for XNode you can test whether LastNode is null.

Comment (JamesNK) [inline Src/Newtonsoft.Json/Converters/XmlNodeConverter.cs]: Same thing with XmlElement.HasAttributes.

Comment (JonHanna) [inline Src/Newtonsoft.Json/Converters/XmlNodeConverter.cs]: Good idea. I'm not sure it's worth doing for attributes though, as it being not-null but empty requires hitting the same collection anyway so it's a bit swing and roundabouts, and it being null is already tested for.

Comment (JonHanna) [inline Src/Newtonsoft.Json/Converters/XmlNodeConverter.cs]: Yes, I tried doing a `HasAttributes` but as well as it being guarded against for the null case already it doesn't offer as much of an improvement. The `HasChildNodes` test is in now, though.

**Comment by JamesNK** [comment]
I thought about it some more after the review and I wrote this - https://github.com/JamesNK/Newtonsoft.Json/commit/96fc838858096a7be4167b8d0deeaf8803ffc0d3

It keeps all the tests for HasChildNodes and HasAttributes internal to the wrappers. XmlNodeConverter will happily use the collections.

If you remove the last commit then I'll squash the others then add mine.

**Comment by JonHanna** [comment]
That's out. I added two comments on your commit (one minor but does have an impact, one trivial).
